Client/Frontend/Browser/Playlist/Cache/PlaylistDownloadManager.swift

```diff
@@ -0,0 +1,485 @@
+        guard let asset = activeDownloadTasks[aggregateAssetDownloadTask] else { return }
+        aggregateAssetDownloadTask.taskDescription = asset.id
+        aggregateAssetDownloadTask.resume()
+        
+        DispatchQueue.main.async {
+            self.delegate?.onDownloadStateChanged(id: asset.id, state: .inProgress, displayName: self.delegate?.displayNames(for: mediaSelection) ?? "")
+        }
+    }
+    
+    func urlSession(_ session: URLSession, aggregateAssetDownloadTask: AVAggregateAssetDownloadTask, didLoad timeRange: CMTimeRange, totalTimeRangesLoaded loadedTimeRanges: [NSValue], timeRangeExpectedToLoad: CMTimeRange, for mediaSelection: AVMediaSelection) {
+        
+        guard let asset = activeDownloadTasks[aggregateAssetDownloadTask] else { return }
+
+        var percentComplete = 0.0
+        for value in loadedTimeRanges {
+            let loadedTimeRange: CMTimeRange = value.timeRangeValue
+            percentComplete +=
+                loadedTimeRange.duration.seconds / timeRangeExpectedToLoad.duration.seconds
+        }
+        
+        DispatchQueue.main.async {
+            self.delegate?.onDownloadProgressUpdate(id: asset.id, percentComplete: percentComplete * 100.0)
+        }
+    }
+}
+
+private class PlaylistFileDownloadManager: NSObject, URLSessionDownloadDelegate {
+    private var activeDownloadTasks = [URLSessionTask: MediaDownloadTask]()
+    
+    weak var delegate: PlaylistStreamDownloadManagerDelegate?
+    
+    func restoreSession(_ session: URLSession, completion: @escaping () -> Void) {
+        session.getAllTasks { [weak self] tasks in
+            guard let self = self else { return }
+            
+            for task in tasks {
+                guard let pageSrc = task.taskDescription else { break }
+                
+                if let item = Playlist.shared.getItem(pageSrc: pageSrc),
+                   let mediaSrc = item.mediaSrc,
+                   let assetUrl = URL(string: mediaSrc) {
+                    let info = PlaylistInfo(item: item)
+                    let asset = MediaDownloadTask(id: info.pageSrc, name: info.name, asset: AVURLAsset(url: assetUrl))
+                    self.activeDownloadTasks[task] = asset
+                }
+            }
+            
+            DispatchQueue.main.async {
+                completion()
+            }
+        }
+    }
+    
+    func downloadAsset(_ session: URLSession, assetUrl: URL, for item: PlaylistInfo) {
+        let asset = AVURLAsset(url: assetUrl)
+        
+        let request: URLRequest = {
+            var request = URLRequest(url: assetUrl, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10.0)
+            request.addValue("bytes=0-", forHTTPHeaderField: "Range")
+            request.addValue(UUID().uuidString, forHTTPHeaderField: "X-Playback-Session-Id")
+            request.addValue("AppleCoreMedia/1.0.0.17E255 (iPhone; U; CPU OS 13_4 like Mac OS X; en_ca)", forHTTPHeaderField: "User-Agent")
```

Comment:
Will this have to update over time? Perhaps it'd be good to keep it somewhere like the main UA we use for Brave?